### PR TITLE
feat: improved RPC's response error handling with HTTP status codes

### LINKED ISSUE #11792: Consider returning http errors for certain rpc errors
Currently when a rpc request is processed, we [never return a http error](https://github.com/near/nearcore/blob/6f321bdd0f2f568a90d764296469aee27a9f8208/chain/jsonrpc/src/lib.rs#L293). However, this makes it confusing to rpc operators, who may see very low "error rate" in their monitoring when there may actually be a nontrivial amount of requests that return a timeout error with a 200 response code. We may want to consider returning a `HttpError` in that function for certain rpc errors like timeout. Of course there are legitimate errors such as "account not found" that make sense to be included in a response body with a 200.

### PR DESCRIPTION
**Pull request by frolvanya**
This PR updates the RPC handler to return appropriate HTTP errors for specific RPC issues, improving error visibility for operators

closes #11792

### CODE REVIEW

File: chain/jsonrpc/src/lib.rs
Comment (shreyan-gupta) [inline chain/jsonrpc/src/lib.rs:291]: Sorry, I'm not super familiar with the codebase in this area, but when I was taking a look at the `Message` enum, it does seem to have a function to convert RpcError to a response. Is that what we might want to use here, maybe just return the `Message` with RpcError instead of `Result<Message, RpcError>`?

Or maybe better, since we know we always expect a "response" out of this function, why not return an object of type `Response` defined [here](https://github.com/near/nearcore/blob/fe4ac1a316cf09fa391636d8532192239f8b5a1c/chain/jsonrpc-primitives/src/message.rs#L76)?

Comment (shreyan-gupta) [reply chain/jsonrpc/src/lib.rs:291]: Additionally, I see this function has only one call site which is in the same file, might be good to get rid of the `pub` here.

File: chain/jsonrpc/src/lib.rs
Comment (shreyan-gupta) [inline chain/jsonrpc/src/lib.rs:1362]: To me it seems like we would never really return a HttpError anyway as `handler.process` never returns an error.

Sorry, could you help me understand the high level motive here?

Comment (shreyan-gupta) [reply chain/jsonrpc/src/lib.rs:1362]: Oh I see, this is supposed to fix the original issue by Bowen. I think what we might want to do here is return the `Response` from `handler.process` and parse the response `result` field to see whether it's a `Value` or `RpcError`. In case of RpcError, convert to HttpError and return that?

**Comment by frolvanya** [comment]
On the one hand, this makes sense. However, If we will send a structure other than Result<Message>, then we will have to rewrite jsonrpc tests. Is it worth it?
image

Also, we need to make `Version` struct `pub`, it's not necessarily a bad idea, but it's definetely a small overhead
image

And process function in the very end will have these ugly raw response builders:
image

**Comment by frolvanya** [comment]
@shreyan-gupta Sorry, it took a bit longer than expected, but I removed Result from process function

**Review by shreyan-gupta — APPROVED** [review]

File: chain/jsonrpc/src/lib.rs
Comment (shreyan-gupta) [inline chain/jsonrpc/src/lib.rs:294]: nit: These are equivalent

```suggestion
            Message::Request(request) => Message::response(id, self.process_request(request).await)
```

Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs]: Thanks! Sorry, I've missed it

**Comment by frolvanya** [comment]
@race-of-sloths please, include my PR in the Race

**Comment by race-of-sloths** [comment]
@frolvanya Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!


[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/frolvanya?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/frolvanya?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/frolvanya)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/frolvanya)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: chain/jsonrpc/src/lib.rs
Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs:1373]: This might work, but do you want me to get rid of async move and return a simple request from this function?

P.S: No, I'm not changing the JSON response. I'm just modifying the status code of the response

Comment (frol) [reply chain/jsonrpc/src/lib.rs:1373]: > do you want me to get rid of async move and return a simple HttpResponse from this function?

I did not ask for that, but if that will work, I would try that. I don't recall the reason for this manual Future boxing (maybe check git blame first to see the reason for this instead of simple `async fn`). If the code has been there for 3+ years, it could be simply due to no async support in Rust back in those days, and so we can change it now.

P.S. I only discovered this PR due to @race-of-sloths ;-)

Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs]: Done. I've simplified `rpc_handler` function and replaced `async move` with a simple `async fn`. Last commit related to `rpc_handler` function header was done 5 years ago by Illia, so as you said Rust was lacking some core async features back then

**Review by khorolets — APPROVED** [review]
LGTM! That has to be done long time ago, thank you!

**Comment by race-of-sloths** [comment]
### 🔄 The PR has been merged.
>[!IMPORTANT]
This pull request is a part of the Race of Sloths and has not been scored yet. Scoring will close in 24 hours! 🕰️

File: chain/jsonrpc/src/lib.rs
Comment (frol) [inline chain/jsonrpc/src/lib.rs:1363]: @frolvanya @khorolets I have just realized that it turns HandlerError into 500 errors, which is not the right thing to do. Handler errors should be HTTP 200 as execution was totally fine and it is just an application level result, it is not an RPC error

Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs:1363]: Do you want me to open a new PR for this?

Comment (frol) [reply chain/jsonrpc/src/lib.rs:1363]: @frolvanya yes, please. Let's align on the types of errors matching the HTTP status codes: https://github.com/near/nearcore/issues/11792#issuecomment-2241189863

Comment (frolvanya) [reply chain/jsonrpc/src/lib.rs:1363]: @frol can you please check if the new PR is your desired fix or not?

**Comment by race-of-sloths** [comment]
### ✅ PR is finalized! 
Your contribution is much appreciated with a final score of 0!
You have received 0 Sloth points for this contribution

**Pull request merged**
